dotnet/sdk-phase-2.md

```diff
@@ -0,0 +1,484 @@
+    methods
+  * We could have returned/accepted null for unset, but this approach is more .NET like
+  * We currently have no public way to get all signal/query handlers from inside a workflow (same as Python)
+* There's no value in accepting workflow type in `GetExternalWorkflowHandle` at this time because we can only signal and
+  cancel it currently. And signal is not currently scoped to the workflow at compile time.
+* We expect users to `throw new ContinueAsNewException()` instead of have some helper
+  * This is the .NET preferred approach (same for `throw new CompleteAsyncException()` for activities)
+* We may add helpers for getting individual memo values given a type to deserialize to
+  * We don't expose a data converter to the user in a workflow at this time, so it is wise to make these helpers
+* A `CancellationToken` can be provided on every async call made, but if not set, by default workflow cancellation token
+  is the token used. This allows it to be overridden with a token not related to workflow cancellation.
+
+TODO(cretz): For discussion:
+
+* `ChildWorkflowHandle` has `GetResultAsync()` to match `WorkflowHandle` in client, but this could just be a `Task`
+  property. Should it be? Other .NET APIs tend to do this.
+* In client I call options `WorkflowStartOptions`, not only for `StartWorkflowAsync` but also for
+  `ExecuteWorkflowAsync`. I have a similar pattern here with activities and children. Should I continue this or should I
+  remove the `Start` word?
+* Should I rename `Now` to `UtcNow` to match .NET since it's non-local? Or does matching `DateTime.UtcNow` not matter
+  that much?
+
+## Workflow Replayer API
+
+```csharp
+namespace Temporalio.Worker;
+
+public class WorkflowReplayer
+{
+    public WorkflowReplayer(WorkflowReplayerOptions options) { /*...*/ }
+
+    public Task<WorkflowReplayResult> ReplayWorkflowAsync(
+        WorkflowHistory history, WorkflowReplayOptions? options = null) { /*...*/ }
+}
+
+public class WorkflowReplayOptions : ICloneable
+{
+    public bool ThrowOnReplayFailure { get; set; }
+}
+
+public record WorkflowReplayResult(Exception? ReplayFailure = null);
+```
+
+Notes about the above code:
+
+* We call this `WorkflowReplayer` instead of `Replayer` like we do in Python
+  * TODO(cretz): Should we call this `Replayer` instead? I think I still like `ReplayWorkflowAsync` instead of just
+    `ReplayAsync` regardless of class name.
+* `WorkflowReplayerOptions` shares many options with `TemporalWorkerOptions`, but not enough (and a few extra from
+  client) to justify any code sharing
+* `ReplayWorkflowAsync` is thread safe and we'll document this instead of making an overload of this that accepts an
+  iterator
+* Unlike Python we choose not to provide the history on the `WorkflowReplayResult` because we don't aggregate for the
+  user (they have LINQ ways to run many in parallel and they can correlate as desired)
```

Comment:
As long as the experience without accepting an iterator is nice, I'm all of less APIs.